### PR TITLE
screencast: Don't cast remote desktop sessions to screencast

### PR DESCRIPTION
**Pull request by GeorgesStavracas**
The replace_restore_token_with_data() function is called both for
remote desktop and screencast sessions. However, it was always
casting and using to ScreenCastSession, even when a remote desktop
session was passed. This could lead to memory corruption.

Be more careful when casting the Session object to ScreenCastSession.
Move the casting entirely to inside replace_restore_token_with_data(),
and make sure the only code paths that perform type casting are safely
guarded by either is_remote_desktop_session or is_screen_cast_session
checks.

This is a regression from the screencast restore PR.

Closes https://github.com/flatpak/xdg-desktop-portal/issues/771

### CODE REVIEW

**Comment by matthiasclasen** [comment]
Good catch.
Looks right to me.

File: src/screen-cast.c [hunk 1]
```diff
@@ -656,6 +663,7 @@ replace_restore_token_with_data (ScreenCastSession *screen_cast_session,
            * apps can pass random UUIDs and yet predict what the next token will
            * be.
            */
+          screen_cast_session = (ScreenCastSession *)session;
```
[the remark this answers is not in the document]
Comment (GeorgesStavracas) [reply src/screen-cast.c:666]: There's an early return above that, but sure

[the remark this answers is not in the document]
Comment (GeorgesStavracas) [reply src/screen-cast.c:667]: Done - you ask for an assert, you get an assert!

**Comment by GeorgesStavracas** [comment]
I think this PR is backport material, by the way. It's worth a release, I'd say.

File: src/screen-cast.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (mwleeds) [reply src/screen-cast.c:667]: There's an early return for `if (is_remote_desktop_session (session))` but I guess maybe we could add a new type of session in the future. Anyway, looks good now

**Comment by GeorgesStavracas** [comment]
Cherry-picked to `xdg-desktop-portal-1.12` and `xdg-desktop-portal-1.14`

**Pull request merged**